From f7382b1180d3f3dc35ddcab4e7d1dbd1652d0f0f Mon Sep 17 00:00:00 2001
From: Hendrik Brueckner <brueckner@redhat.com>
Date: Thu, 21 Aug 2014 14:48:52 -0400
Subject: [s390] dasd: fix list_del corruption during format

Message-id: <1408632535-642-4-git-send-email-brueckner@redhat.com>
Patchwork-id: 88681
O-Subject: [RHEL7.1 PATCH] [s390] dasd: fix list_del corruption during format
Bugzilla: 1132371
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Description
-----------
commit 8fa56aed12f1b0a2828da52280e2efbbf1163ad5
Author: Stefan Haberland <stefan.haberland@de.ibm.com>
Date:   Fri Jul 18 14:24:20 2014 +0200

    dasd: fix list_del corruption during format

    If I/O errors occur during format a kernel panic with a list_del
    corruption may occur.

    Stop error recovery procedure after an erp action was taken.

    Signed-off-by: Stefan Haberland <stefan.haberland@de.ibm.com>
    Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>

Bugzilla
--------
BZ 1132371
https://bugzilla.redhat.com/show_bug.cgi?id=1132371

Upstream status of the patch
----------------------------
The patch is upstream as of kernel version
http://git.kernel.org/?p=linux/kernel/git/torvalds/linux-2.6.git;a=commitdiff;h=8fa56aed12f1b0a2828da52280e2efbbf1163ad5

Brew
----
https://brewweb.devel.redhat.com/taskinfo?taskID=7869031

Test status
-----------
The patch has been tested and fixes the problem.
The fix has been verified by the IBM test department.

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/s390/block/dasd.c b/drivers/s390/block/dasd.c
index 523be1c..242fa98 100644
--- a/drivers/s390/block/dasd.c
+++ b/drivers/s390/block/dasd.c
@@ -2293,8 +2293,12 @@ retry:
 
  rc = 0;
  list_for_each_entry_safe(cqr, n, ccw_queue, blocklist) {
-  if (__dasd_sleep_on_erp(cqr))
+  if (__dasd_sleep_on_erp(cqr)) {
+   if (!cqr->status == DASD_CQR_TERMINATED &&
+       !cqr->status == DASD_CQR_NEED_ERP)
+    break;
    rc = 1;
+  }
  }
  if (rc)
   goto retry;
-- 
1.7.1